File: leetcode/python/Copy List with Random Pointer/Copy_List_with_Random_Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if head is None:
            return None
        h = head
        p = h
        #第一遍扫描：在原表后添加新的节点
        while p is not None:
            next_node = p.next
            new_node = RandomListNode(p.label)
            p.next = new_node
            new_node.next = next_node
            p = next_node
        p = h
        #第二遍扫描：新节点添加随机指针
        while p is not None:
            next_node = p.next.next;
            new_node = p.next
            if p.random is not None:
                new_node.random = p.random.next
            p = next_node

        
        #第三遍扫描：将新节点从原来的表中拆下来
        res = None
        end = None
        p = h
        while p is not None:
            next_node = p.next.next
            new_node = p.next
            if res is None:
                res = new_node
                end = new_node
            else:
                end.next = new_node
                end = end.next
            p.next = next_node
            p = next_node
        return res

File: leetcode/cpp/CopyListwithRandomPointer.cpp
//
// Created by zhangqi on 2017/10/31.
// Time:  O(n)
// Space: O(1)

    struct RandomListNode {
         int label;
         RandomListNode *next, *random;
         RandomListNode(int x) : label(x), next(nullptr), random(nullptr) {}
     };

class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        //在每个结点后创建一个新结点
        for(RandomListNode *curr = head;curr; curr = curr->next->next){
            RandomListNode *node = new RandomListNode(curr->label);
            node->next = curr->next;
            curr->next = node;
        }

        //将新结点指向正确的位置
        for(auto curr = head; curr; curr = curr->next->next){
            if(curr->random){
                curr->next->random = curr->random->next;
            }
        }

        //分开两个链表
        RandomListNode dummy{0};

        for(auto curr = head ,copyNode = &dummy; curr; ){
            copyNode->next = curr->next;
            curr->next = curr->next->next;

            curr = curr->next;
            copyNode = copyNode->next;
        }

        return  dummy.next;

    }
};